ambry-protocol/src/main/java/com.github.ambry.protocol/AdminRequest.java

```diff
@@ -0,0 +1,149 @@
+  private final long sizeInBytes;
+
+  private int sizeSent = 0;
+
+  /**
+   * Reads from a stream and constructs an {@link AdminRequest}.
+   * @param stream the {@link DataInputStream} to read from.
+   * @param clusterMap the {@link ClusterMap} in use.
+   * @return {@link AdminRequest} that is deserialized from the {@code stream}.
+   * @throws IOException if there is an I/O error reading from {@code stream}
+   */
+  public static AdminRequest readFrom(DataInputStream stream, ClusterMap clusterMap) throws IOException {
+    Short versionId = stream.readShort();
+    int correlationId = stream.readInt();
+    String clientId = Utils.readIntString(stream);
+    AdminRequestOrResponseType type = AdminRequestOrResponseType.values()[stream.readShort()];
+    AdminRequest request;
+    switch (type) {
+      case TriggerCompaction:
+        PartitionId id = clusterMap.getPartitionIdFromStream(stream);
+        request = new AdminRequest(type, id, correlationId, clientId);
+        break;
+      default:
+        throw new IllegalArgumentException("Unrecognized admin request type: " + type);
+    }
+    // ignore version for now
+    return request;
+  }
+
+  /**
+   * Constructs an admin request with the given parameters.
+   * @param type the type of the request.
+   * @param partitionId the {@link PartitionId} that the operation should work on. {@link null} if not applicable.
+   * @param correlationId an ID to help match responses to requests.
+   * @param clientId the ID of the client.
+   */
+  public AdminRequest(AdminRequestOrResponseType type, PartitionId partitionId, int correlationId, String clientId) {
+    super(RequestOrResponseType.AdminRequest, ADMIN_REQUEST_VERSION_V1, correlationId, clientId);
+    this.type = type;
+    this.partitionId = partitionId;
+    sizeInBytes = computeSizeInBytes();
+  }
+
+  @Override
+  public long writeTo(WritableByteChannel channel) throws IOException {
+    long written = 0;
+    if (bufferToSend == null) {
+      serializeIntoBuffer();
+    }
+    if (bufferToSend.remaining() > 0) {
+      written = channel.write(bufferToSend);
+      sizeSent += written;
+    }
+    return written;
+  }
+
+  @Override
+  public boolean isSendComplete() {
+    return sizeSent == sizeInBytes();
+  }
+
+  @Override
+  public long sizeInBytes() {
+    return sizeInBytes;
+  }
+
+  /**
+   * @return the type of the request
+   */
+  public AdminRequestOrResponseType getType() {
+    return type;
+  }
+
+  /**
+   * @return the {@link PartitionId} that the operation will work on. {@link null} if not applicable.
+   */
+  public PartitionId getPartitionId() {
+    return partitionId;
+  }
+
+  @Override
+  public String toString() {
+    return "AdminRequest[ClientId=" + clientId + ", CorrelationId=" + correlationId + ", Type=" + type
+        + ", PartitionId=" + partitionId + "]";
+  }
+
+  /**
+   * @return the size in bytes of the serialized version of the request
+   */
+  private long computeSizeInBytes() {
```

Comment:
why not have this method set the member variable directly? Alternatively, rename this as `computeAndGetSizeInBytes()`?